### PR TITLE
Benchmarks for Accounts and introduction of bytes pool

### PR DESCRIPTION
**Pull request by JekaMas**
To run benchmarks - `go test -bench=. -benchmem` in `core/types/accounts` directory.

The PR introduces `common/pool` package that can be used to get `[]byte` with zero allocations. In a nutshell, it's a bunch of buckets of `sync.Pool` for different capacity slices.

EncodeForHashing, EncodeForStorage, Decode - a bit optimized by using `Pool` and now have 0 allocations.

DecodeRLP has 3 allocations from reading from `*rlp.Stream` type and 6 more from `big.Int` manipulations.
EncodeRLP - only 6 more from `big.Int` manipulations.

Further optimizations *if we need them*: I think it's possible to use `Pool` at the database level also it's possible to use a more optimized version of `big.Int`.

*I'm going post actual benchmark results tomorrow morning, the tests are still running.

### CODE REVIEW

File: trie/resolver.go [hunk 1]
```diff
@@ -224,7 +225,9 @@ func (tr *TrieResolver) Walker(keyIdx int, k []byte, v []byte) (bool, error) {
 			}
 			encodeLen := tr.a.EncodingLengthForHashing()
 			if int(encodeLen) > len(tr.buf) {
-				tr.buf = make([]byte, encodeLen)
+				pool.PutBuffer(pool.NewValue(tr.buf))
+				buf := pool.GetBuffer(encodeLen)
+				tr.buf = buf.B
```
Comment (AlexeyAkhunov) [inline trie/resolver.go]: You can now remove `buf` field from the type `Resolver` and just use the buffer from the pool?

Comment (JekaMas) [reply trie/resolver.go]: It'd be nice to do. Checking...

**Comment by JekaMas** [comment]
```
BenchmarkEncodingLengthForStorage/EmptyAccount-8                10000000               143 ns/op               8 B/op          0 allocs/op
BenchmarkEncodingLengthForStorage/AccountEncodeWithCode-8               10000000               191 ns/op               8 B/op          0 allocs/op
BenchmarkEncodingLengthForStorage/AccountEncodeWithCodeWithStorageSizeHack-8            10000000               196 ns/op               8 B/op          0 allocs/op
BenchmarkEncodingLengthForStorage/AccountEncodeWithCodeWithStorageSizeEIP2027-8         10000000               193 ns/op               8 B/op          0 allocs/op
BenchmarkEncodingLengthForHashing/EmptyAccount-8                                        10000000               211 ns/op               8 B/op          0 allocs/op
BenchmarkEncodingLengthForHashing/AccountEncodeWithCode-8                               10000000               198 ns/op               8 B/op          0 allocs/op
BenchmarkEncodingLengthForHashing/AccountEncodeWithCodeWithStorageSizeHack-8            10000000               196 ns/op               8 B/op          0 allocs/op
BenchmarkEncodingLengthForHashing/AccountEncodeWithCodeWithStorageSizeEIP2027-8         10000000               194 ns/op               8 B/op          0 allocs/op
BenchmarkEncodingAccountForStorage/EmptyAccount-8                                        5000000               252 ns/op               0 B/op          0 allocs/op
BenchmarkEncodingAccountForStorage/AccountEncodeWithCode-8                               5000000               363 ns/op               0 B/op          0 allocs/op
BenchmarkEncodingAccountForStorage/AccountEncodeWithCodeWithStorageSizeHack-8            5000000               363 ns/op               0 B/op          0 allocs/op
BenchmarkEncodingAccountForStorage/AccountEncodeWithCodeWithStorageSizeEIP2027-8         5000000               364 ns/op               0 B/op          0 allocs/op
BenchmarkEncodingAccountForHashing/EmptyAccount-8                                        5000000               365 ns/op               0 B/op          0 allocs/op
BenchmarkEncodingAccountForHashing/AccountEncodeWithCode-8                               5000000               355 ns/op               0 B/op          0 allocs/op
BenchmarkEncodingAccountForHashing/AccountEncodeWithCodeWithStorageSizeHack-8            5000000               362 ns/op               0 B/op          0 allocs/op
BenchmarkEncodingAccountForHashing/AccountEncodeWithCodeWithStorageSizeEIP2027-8         5000000               363 ns/op               0 B/op          0 allocs/op
BenchmarkDecodingAccount/EmptyAccount-8                                                  2000000               844 ns/op              63 B/op          2 allocs/op
BenchmarkDecodingAccount/AccountEncodeWithCode-8                                         2000000               871 ns/op              63 B/op          2 allocs/op
BenchmarkDecodingAccount/AccountEncodeWithCodeWithStorageSizeHack-8                      2000000               873 ns/op              63 B/op          2 allocs/op
BenchmarkDecodingAccount/AccountEncodeWithCodeWithStorageSizeEIP2027-8                   2000000               874 ns/op              63 B/op          2 allocs/op
BenchmarkRLPEncodingAccount/EmptyAccount-8                                              20000000               115 ns/op               0 B/op          0 allocs/op
BenchmarkRLPEncodingAccount/AccountEncodeWithCode-8                                     20000000                94.7 ns/op             0 B/op          0 allocs/op
BenchmarkRLPEncodingAccount/AccountEncodeWithCodeWithStorageSizeHack-8                  20000000               100 ns/op               0 B/op          0 allocs/op
BenchmarkRLPEncodingAccount/AccountEncodeWithCodeWithStorageSizeEIP2027-8               20000000                96.9 ns/op             0 B/op          0 allocs/op
BenchmarkRLPDecodingAccount/EmptyAccount-8                                               1000000              1269 ns/op             243 B/op          3 allocs/op
BenchmarkRLPDecodingAccount/AccountEncodeWithCode-8                                      1000000              1563 ns/op             310 B/op          6 allocs/op
BenchmarkRLPDecodingAccount/AccountEncodeWithCodeWithStorageSizeHack-8                   1000000              1689 ns/op             314 B/op          6 allocs/op
BenchmarkRLPDecodingAccount/AccountEncodeWithCodeWithStorageSizeEIP2027-8                1000000              1626 ns/op             312 B/op          6 allocs/op
BenchmarkIsEmptyCodeHash-8                                                              300000000                6.09 ns/op            0 B/op          0 allocs/op
BenchmarkIsEmptyRoot-8                                                                  200000000                6.54 ns/op            0 B/op          0 allocs/op
```

File: trie/resolver.go (hunk 1 above)
Comment (JekaMas) [reply trie/resolver.go]: And it works :)

**Pull request merged**